From cad25b3f0660499b8815e545e87d23506d202710 Mon Sep 17 00:00:00 2001
From: Todd Kennedy <toddke@google.com>
Date: Tue, 31 Oct 2017 13:05:55 -0700
Subject: [PATCH 009/306] Set updated system package's mExtra

While scanning system packages, we abort the process in the
middle of the scan and leave our internal structures in an
ill defined state. This change is not about making the structures
consistent; rather unblocking a build breakage. We'll look
at data consistency during a later refactoring stage when
the process of package scanning is fixed.

A reboot is necessary for the bug to exhibit itself because
the permissions are re-granted [or denied] on every boot.

Change-Id: I96cc2f76623911f2cf93727e9f1787b42210a8d6
Fixes: 68328870
Test: Manual.
Test: Add a new priv-app that take a privieleged permission.
Test: See that the permission is granted.
Test: Update the app and reboot
Test: See that the permission is still granted.
Test: cts-tradefed run commandAndExit cts-dev -m CtsAppSecurityHostTestCases -t android.appsecurity.cts.PermissionsHostTest
Test: cts-tradefed run commandAndExit cts-dev -m CtsPermissionTestCases
Test: cts-tradefed run commandAndExit cts-dev -m CtsPermission2TestCases
Test: bit FrameworksServicesTests:com.android.server.pm.PackageManagerSettingsTests
(cherry picked from commit 5f2faaa7ed71a3d1aa862c37e27ebf817f094dc5)
---
 .../core/java/com/android/server/pm/PackageManagerService.java   | 1 +
 1 file changed, 1 insertion(+)

diff --git a/services/core/java/com/android/server/pm/PackageManagerService.java b/services/core/java/com/android/server/pm/PackageManagerService.java
index 7be0cde4a94..7837b029830 100644
--- a/services/core/java/com/android/server/pm/PackageManagerService.java
+++ b/services/core/java/com/android/server/pm/PackageManagerService.java
@@ -8687,6 +8687,7 @@ public class PackageManagerService extends IPackageManager.Stub
                 pkg.applicationInfo.primaryCpuAbi = updatedPkg.primaryCpuAbiString;
                 pkg.applicationInfo.secondaryCpuAbi = updatedPkg.secondaryCpuAbiString;
             }
+            pkg.mExtras = updatedPkg;
 
             throw new PackageManagerException(Log.WARN, "Package " + ps.name + " at "
                     + scanFile + " ignored: updated version " + ps.versionCode
-- 
2.17.1

